tools/build_script_generator/scons/site_tools/gdb.py

```diff
@@ -23,21 +23,22 @@
 # Start the gnu debugger
 # -----------------------------------------------------------------------------
 
-def gdb_debug(env, source, alias='gdb_debug'):
-	actionString = '$GDB $SOURCE -ex "target extended-remote :$GDB_PORT"'
-	action = Action(actionString, cmdstr="$GDB_COMSTR")
-	return env.AlwaysBuild(env.Alias(alias, source, action))
+def gdb_openocd_debug(env, source, alias='gdb_openocd_debug'):
+
+
+	# pre = Action("openocd")
+	action = Action("openocd & arm-none-eabi-gdb; kill $!", cmdstr="$GDB_COMSTR")
+	target = env.Alias(alias, source, action)
+	# AddPreAction(target, pre)
+	return env.AlwaysBuild(target)
```

Comment:
Anyone got an idea how to start OpenOCD in the background before running arm-none-eabi-gdb and then killing it together? This command works fine in my shell, but not in SCons, I assume due to minimal shell environment.
Hey, @dergraaf, you know a little more SCons than me, I tried PreAction, but it's blocking, and `openocd &` doesn't work either.

Reply:
You probablly need to use pure Python (e.g. the subprocess module). SCons does not really support long running tasks which build nothing. Which is fine, as it starts leaving the realms of what a build system should do.

Reply:
> Which is fine, as it starts leaving the realms of what a build system should do.

I actually agree, there quite a lot of modm-specific SCons tools that maybe should be pure python tools instead (like `font`, `bitmap`, `unittest`) and wrapped into SCons tools like `system_design` or integrated into modm itself.

The CMake builder has a Makefile that calls openocd/gdb manually, maybe we should have something similar? But it's getting a little crowded now in the main directory:

- `modm/` from modm
- `SConstruct` for scons
- `openocd.cfg` for openocd
- `.gdbinit` for gdb
- `project.xml.log` from lbuild
- `CMakeLists.txt` for CMake
- `Makefile` CMake wrapper
- `*.creator`, `*.files`, `*.config`, `*includes` for Qt Creator

Reply:
In my current projects I often use a Makefile (which is more of a glorified bash script) to start different tools. Haven't found a better/nicer solution yet.

Reply:
I've still done this with subprocess and in SCons, since it's more streamlined to have the automatic dependency on the debug and release targets for debugging. I still had to use `setsid` in Python to get GDB to not kill OpenOCD, like @daniel-k, but whatevs, now it works.